internal/repository/file.go

```diff
@@ -0,0 +1,295 @@
+package repository
+
+import (
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"fyne.io/fyne"
+	"fyne.io/fyne/storage"
+	"fyne.io/fyne/storage/repository"
+)
+
+// fileSchemePrefix is used for concatenating with paths when we must construct
+// them from strings.
+const fileSchemePrefix string = "file://"
+
+// declare conformance with repository types
+var _ repository.Repository = (*FileRepository)(nil)
+var _ repository.WriteableRepository = (*FileRepository)(nil)
+var _ repository.HierarchicalRepository = (*FileRepository)(nil)
+var _ repository.ListableRepository = (*FileRepository)(nil)
+var _ repository.MovableRepository = (*FileRepository)(nil)
+var _ repository.CopyableRepository = (*FileRepository)(nil)
+
+var _ fyne.URIReadCloser = (*file)(nil)
+var _ fyne.URIWriteCloser = (*file)(nil)
+
+type file struct {
+	*os.File
+	uri fyne.URI
+}
+
+func (f *file) URI() fyne.URI {
+	return f.uri
+}
+
+// FileRepository implements a simple wrapper around golang's filesystem
+// interface libraries. It should be registered by the driver on platforms
+// where it is appropriate to do so.
+//
+// This repository is suitable to handle the file:// scheme.
+//
+// Since: 2.0.0
+type FileRepository struct {
+}
+
+// NewFileRepository creates a new FileRepository instance. It must be
+// given the scheme it is registered for. The caller needs to call
+// repository.Register() with the result of this function.
+//
+// Since: 2.0.0
+func NewFileRepository() *FileRepository {
+	return &FileRepository{}
+}
+
+// Exists implements repository.Repository.Exists
+//
+// Since: 2.0.0
+func (r *FileRepository) Exists(u fyne.URI) (bool, error) {
+	p := u.Path()
+
+	_, err := os.Stat(p)
+	ok := false
+
+	if err == nil {
+		ok = true
+	} else if os.IsNotExist(err) {
+		err = nil
+	}
+
+	return ok, err
+}
+
+func openFile(uri fyne.URI, create bool) (*file, error) {
+	path := uri.Path()
+	var f *os.File
+	var err error
+	if create {
+		f, err = os.Create(path) // If it exists this will truncate which is what we wanted
```

Comment:
This will panic if the parent directories don't exist, and there doesn't seem to be anyway to create directories with this API

Reply:
>  and there doesn't seem to be anyway to create directories with this (URI/Repository) API.

I didn't think of that. That's probably something we should address. But a "directory" only makes sense in the context of `file://`. Other types of repositories may have different types of resources.

Maybe we could add a `CreateListable` to `WriteableRepository()` ? For repositories where they are the same like `InMemoryRepository` it could just be equivalent to `touch`. I think that's a common enough thing to want to do.

Reply:
I think this also raises the question of whether or not we want to expose some mechanism for passing configuration/flags to `Writer()`, `CreateListable()`, etc. The problem is that the options that are available will vary for every storage repository, and may even vary by OS (permissions for `file://` on Windows work much different than on Linux).

I think I would lean towards addressing this by:

* Checking that the parent exists, and returning an error if not.
* Adding `CreateListable()` to `WriteableRepository`.

We can address options/flags in the future.

Reply:
I agree this mainly makes sense in `file://` so I would propose implementing the solution in the link to solve the immediate issue, we can revisit extensions in future.

```go
if err := os.MkdirAll(filepath.Dir(path), 0770); err != nil {
    return nil, err
}
f, err := os.Create(p)
```

Reply:
Please expand the tests to cover this situation;
- Writing to "/" - should not be possible, not even sure how `filepath.Dir` would respond
- Writing to "/a/b" - if "/a" doesn't exist it should be created
- Writing to "/a/c/d/c" - if "/a", "/a/c", and "/a/c/d" don't exist they should all be created

Reply:
I agree that adding `CreateListable` would be useful, and also that creating a resource in a place that does not exist should return an error. I  wonder if the right place for `CreateListable` would be `ListableRepository` rather than `WritableRepository` however.
I don't think we should attempt to automatically create the parents if that type of error would have occurred - it feels like we are making many assumptions for the developer.

Reply:
Maybe we could have a `ListWriter` type which has `CreateListable`? I agree with @andydotxyz that we should not create parent directories.

Writing to `/` should fail anyway, because it is a directory, so the underlying call into the fs should fail.

Reply:
I'm fine with a more explicit way to create directories, rather than hidden assumptions.

I am not sure of the best name for this operation, nor the best location for it, tho I agree `WritableRepository.CreateXYZ` doesn't feel like the right place, so that leaves either `ListableRepository.CreateListable` or `HierarchicalRepository.CreateParent`.

Conceptually there is an overlay between `ListableRepository` and `HierarchicalRepository` as they both deal with hierarchy (parent/child relationships).

Reply:
> Writing to / should fail anyway, because it is a directory, so the underlying call into the fs should fail.

👍

Reply:
Something can be listable without having children though. I think I am satisfied with the Hierarchical/Listable split. I think `CreateListable` should go in `ListableRepository`. If the repository is read only, it can return `ErrOperationNotSupported`.

Or like I said, we could create a dedicated interface for this, something like `ListableCreatorRepository`. That might be the more elegant solution.

Reply:
Either way how would this be surfaced to Fyne developers? ie. if a developer wants to create a directory are they expected to look up the repository, type cast it, and call the operation, or would there be something like `storage.XYZ`?

Reply:
There would be a `storage.CreateListable()`

Reply:
My personal preference would be ListableRepository.CreateListable over a new repository just for this 1 method. Though we really are down to the wire to add to these interfaces before they are set - can we resolve this and land tonight?

Reply:
`repository.ListableRepository.CreateListable` and `storage.CreateListable` 👍

Reply:
I have done this in fa994f811d5792f72d5ef4cf1169bb1a7a68fe44. Take a look at the API. I will create tests presently in a subsequent commit.

Reply:
Tests added in 03aca1248060a87f8c057db5c3c84887606f5021. Anybody have issues with this as implemented? If not, I will document the decision and add it to the list of divergences from the prop.

Reply:
Also, to address @stuartmscott 's original issue, I have added a test to ensure this does not panic, and does throw an error in 635f0bf2c27b2522d4a097c6024513c35a215c46.